### PR TITLE
Update SHOW DATABASES to read from mz_catalog.mz_databases

### PR DESCRIPTION
**Pull request by JLDLaughlin**
Part of this larger effort: https://github.com/MaterializeInc/materialize/issues/2157.

Updates `SHOW DATABASES` to read from the `mz_catalog.mz_databases` relation instead of calling the `list_databases` function on the catalog.

Note: This PR also renames the column output of `SHOW DATABASES` from `Database` to `database`. Although this no longer mirrors MySQL, it means that we can run a `SHOW DATABASES` query like:
```
show databases where database = 'materialize';
```
Instead of:
```
show databases where "Database" = 'materialize';
```

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4033)

### CODE REVIEW

File: src/sql/src/plan/statement.rs [hunk 1]
```diff
@@ -446,13 +460,28 @@ fn handle_show_databases(
...
+                    ColumnType {
+                        nullable: false,
+                        scalar_type: ScalarType::String,
+                    },
+                ],
+                keys: vec![],
+            },
+        }),
+        outputs: vec![1], // Only return the database name, not the id.
+    };
```
Comment (benesch) [inline src/sql/src/plan/statement.rs:483]: The amount of hardcoding here of magic numbers and types is worrying to me. What if instead you synthesized a `sql::ast::Select` that specified `SELECT database FROM mz_catalog.mz_databases WHERE ...`?

There's an example of building up `Select` statements programmatically here: https://github.com/MaterializeInc/materialize/blob/be9d85fabb7a2713b6426b1482ec0901f3671d76/src/sql/src/plan/transform_ast.rs#L324-L350

You can add more builder helper methods to the AST types as well!

Comment (JLDLaughlin) [reply src/sql/src/plan/statement.rs]: I think [this commit](https://github.com/MaterializeInc/materialize/pull/4033/commits/4b69b38cad63348b1124cb2c645d58f697fe04be) does what you mean!

File: src/sql-parser/src/parser.rs [hunk 2]
```diff
@@ -2532,7 +2532,21 @@ impl Parser {
...
+                    relation: TableFactor::Table {
+                        name: ObjectName(vec![Ident::new("mz_databases")]),
+                        alias: None,
+                    },
+                    joins: vec![],
+                })
+                .project(SelectItem::Expr {
+                    expr: Expr::Identifier(vec![Ident::new("database")]),
+                    alias: None,
+                });
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:2546]: This snippet looks great, but putting it in the parser is a bit weird! It couples the SQL parser to the details of the `mz_catalog` views, which isn't a coupling that exists before this commit. But I think it should be super straightforward to just leave the AST as-is but move this bit to the `handle_show_databases` function in the planner?

File: src/sql/src/plan/statement.rs [hunk 3]
```diff
@@ -444,15 +458,10 @@ fn finish_show_where(
...
+    ShowDatabasesStatement { query, filter }: ShowDatabasesStatement,
 ) -> Result<Plan, anyhow::Error> {
-    let rows = scx
-        .catalog
-        .list_databases()
-        .map(|database| vec![Datum::from(database)])
-        .collect();
-
-    finish_show_where(scx, filter, rows, &SHOW_DATABASES_DESC)
+    let (expr, _, _) = query::plan_root_query(scx, *query, QueryLifetime::OneShot)?;
```
Comment (benesch) [inline src/sql/src/plan/statement.rs:463]: Following on from above comment, if you can stick the `filter` here onto the `Select` that you synthesize, then you should be able to let `plan_root_query` handle the filtering too. In fact, I think you'd probably just want to call `handle_select` on the synthesized `Select`!

Sorry—maybe I've been a bit obtuse in describing the motivations!—but when everything has been ported over to `mz_catalog` views, we should be able to ditch `plan_show_where` and `finish_show_where`, and just let the normal `handle_select` code path handle everything!

File: src/sql-parser/src/parser.rs [hunk 4]
```diff
@@ -2532,9 +2532,12 @@ impl Parser {
 
     fn parse_show(&mut self) -> Result<Statement, ParserError> {
         if self.parse_keyword("DATABASES") {
-            return Ok(Statement::ShowDatabases(ShowDatabasesStatement {
-                filter: self.parse_show_statement_filter()?,
-            }));
+            let filter = if self.parse_keyword("WHERE") || self.parse_keyword("LIKE") {
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:2535]: This doesn't seem like it handles `LIKE` right? Like, probably you need to leave the AST as is and unpack the `ShowStatementFilter` into the appropriate `WHERE` clause in the sql crate? But this seems well tested, so as long as those tests are passing I'm good with it!

File: src/sql/src/plan/statement.rs [hunk 5]
```diff
@@ -446,13 +447,30 @@ fn handle_show_databases(
...
+        scx,
+        SelectStatement {
+            query: Box::new(Query::select(select)),
+            as_of: None,
+        },
+        &Params {
+            datums: Row::pack(&[]),
+            types: vec![],
+        },
+    )
```
Comment (benesch) [inline src/sql/src/plan/statement.rs:473]: Brilliant! You'll probably find as you continue to port things over that the guts of this function get repeated a bunch—up to you to figure out if you want to share some code + how. (But I think spelling it all out explicitly like this while there is only one example is 100% the way to go.)

**Comment by benesch** [comment]
This turned out really nicely!

**Pull request merged**